File: 12-18-2024/Maximum_Subarray.java
package practice;

public class Maximum_Subarray {
    public static int maxSubArray(int[] nums) {
        int maxSum = Integer.MIN_VALUE, curSum = 0;

        for (int num : nums) {
            curSum += num;
            maxSum = Math.max(curSum, maxSum);
            curSum = Math.max(curSum, 0);
        }

         return maxSum;
    }

    public static void main(String[] args) {
        int[] arr = new int[]{-1};
        System.out.println(maxSubArray(arr));
    }
}

File: 2026-03-28/Maximum_Subarray.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0

        sm = 0
        mx_sum = -1 * (1 << 31)

        for num in nums:
            sm += num
            mx_sum = max(mx_sum, sm)

            if sm < 0:
                sm = 0

        return mx_sum


sol = Solution()
print(sol.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))